Added bridge message templating support

IncognitoCoding opened the pull request:
Yesterday, I added my NetApp storage array through Prometheus for metrics/alerting and realized the sending JSON would not be limited in data because of the fields used, and I would have no easy way of formatting the message since the alerts can not be set up through Grafana.

This got me thinking about how I could implement file templating into the bridge, and once I figured out a plan, I started drafting up the code. 

This PR will allow dedicated user-defined templates based on the Gotify software token. This feature can enable alerts not sent from Grafana to have the ability to present to Gotify cleanly using any `key` value from the JSON.

I had to shuffle around some of the code with a few if/else statements so that you will see a decent chunk of red, but a good amount of the code is just being shifted by `tab`. 

I added several checks to the code and set it up to prevent user-defined templates from crashing the bridge. Anything not done correctly in the templates will have the bridge revert the message to default alerting.

I did not update the docker-compose `ReadMe` section for the bind mount because I did not know how you wanted to add that into the docker, so I left it open for your edits.

Let me know what you think.

IncognitoCoding committed:
Added bridge message templating support

IncognitoCoding committed:
Removed an unneeded bool return and renamed a function.

1. (Removed Bool) The filematch bool return was not required when no file can be checked with a nil tmpls variable value.
2. (Function Rename) Renamed the createMsgTemplate to createUserTemplate for a cleaner name.

DRuggeri commented on main.go:
```diff
@@ -461,11 +553,85 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
 	http.Error(w, strings.Join(text, "\n"), respCode)
 }
 
+func parseMsgTemplates(tmplPath string, token string) (*ut.Template, error) {
```
This feels heavyweight to do in the context of a request. Since it doesn't look like the token string is used, couldn't this walking and gathering of templates be done during startup? This would allow bailing out on error instead of failing all templates of a certain file extension.

IncognitoCoding replied:
This was by design after some extended testing. A ParseGlob failure while parsing the templates will fail (no parse) all files with that extension if one of the templates fails. No templates with that extension will return, but templates with other extensions without fail will return. 

Sorry, the "token" string is no longer used in this function. I mean to remove it after my last revision. I was using the token to try and come up with a workaround for the above statement on the extension failure. The token is not required while using the ParseGlob because the ParseGlob will take into account all user templates, including storing a reference for any defined template tokens. The ParseGlob takes care of all the back-end parsing, and all that needs to be used to use the templates is when using the "ExectueTemplate" you add in the token. Since the ParseGlob parses and stores everything in "tmpls", no custom maps are required.

IncognitoCoding replied:
Also, the `parseMsgTemplates` could be added during startup instead of the handle since the token is not required as part of the call. I had it in the `handleCall` to allow the user templates to be dynamically updated without any restart of the bridge.

DRuggeri replied:
Aye - thank you for the response. I think moving this into startup would address my primary concern (which I didn't quite explain well). I'm thinking that: as a user of the software, if my templates are bad I would want to know that ASAP. It would be bad if my alert gets lost due to something I could have corrected as I deployed my templates, so moving this error detection early in the startup cycle would help prevent that scenario.

(for reference, my deployment workflow is via Chef where the binary, systemd unit, config files, and service start/restart logic are all managed together - but those running in containers/K8S would also benefit from an init failure as a problem signal)

DRuggeri commented on README.md:
```diff
@@ -182,6 +182,100 @@ curl http://127.0.0.1:8080/gotify_webhook -d '
+- The default directory for all templates is the root of the bridge in the folder called `templates`.
+- User-defined templating allows matching and linking using the "define" name of the template [Go Templating](https://golang.google.cn/pkg/text/template/).
+- The Gotify software token is used for matching a message template.
+- The Gotify software token with `title=` at the beginning is used for matching a title template.
```
Why? This seems to rely on magic strings which can often lead to confusion. Could it be possible to instead expect files of a specific name such as tokenfoo.title.tmpl and tokenfoo.message.tmpl? This would allow an inexpensive check during request time to know if a title or message template is used.

IncognitoCoding replied:
When working with "defined" templates, something within the "defined" name must be set to determine a match when executing a template. This is the best option when using a more open/unrestricted folder structure because no file name reads are required. We could add a check, but the detected templates with a title would need to be tracked so know the defined template with the token is a title vs. a message.

DRuggeri replied:
OK, I think I follow. I'm still fairly new to the world of templates (I know how to update a single template that already exists or to operate on a single template, but the rest requires spelunking source code to wrap my head around it). Maybe the wording of the statement just threw me off. Would it be better to maybe say something along the lines of "If a template is defined as `title=TOKEN`, it will be used to render the title of the alert" (depending how the folder structure convo pans out)

IncognitoCoding replied:
Yeah, I think that sounds better. If you were a little thrown off, it would have also caused it to others. I am submitting that ReadMe revision now.

DRuggeri commented on README.md:
```diff
@@ -182,6 +182,100 @@ curl http://127.0.0.1:8080/gotify_webhook -d '
+- The Gotify software token with `title=` at the beginning is used for matching a title template.
+- All file names must be unique but can be any name or subfolder naming you choose.
+- Only one Gotify software token should be defined. More than one will result in inconsistent alerting results.
+- The bridge supports the following template file extensions: "html", "gohtml", "gotmpl", and "tmpl".
```
To keep it simple, I would suggest just using `tmpl`. Specifically, I would not suggest `html`-based extensions to make it clear that html sanitization and escaping will not be done.

IncognitoCoding replied:
Good point! We could certainly remove HTML.

DRuggeri commented on main.go:
```diff
@@ -461,11 +553,85 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
+	for _, p := range fileExt {
+		matchedTmpls, err := ut.ParseGlob(tmplPath + "/*." + p)
+		if err == nil {
+			tmpl = ut.Must(matchedTmpls, err)
```
At first I was going to say that we shouldn't use Must here since it will lead to a panic if there are errors, but then I noticed that the line before this ensures there are no errors... so is this line needed at all?

IncognitoCoding replied:
Yes, the two different `must` lines are required when parsing through root/subdirectories.

DRuggeri replied:
OK, thanks. I asked because [the source for Must](https://cs.opensource.google/go/go/+/refs/tags/go1.20.2:src/text/template/helper.go;l=24-29) just panics in case the err is not `nil` (which we guard against the line immediately before it). So, I think this is just a duplicate check.

I didn't elaborate well on why it was a concern. We should avoid panics during HTTP request servicing if possible. I can't remember if the panic gets `recover`ed by the servemux and just turns into an HTTP 500 to the user or if it actually crashes the whole process. I do recall that the failure never makes it to the log stream unless one writes some recovery middleware to catch it. So, in those scenarios (particularly for function calls), it's usually better to add error handling around the function call. In this case, we already have that since the second return of the function is an `error` and we're paying attention to it in the caller.

IncognitoCoding replied:
As you read through that section, I think you are assuming the first `ParseGlob`/`Must` would parse all templates through the entire file/folder structure. When I started working with templates, I assumed the same thing.

The first, `ParseGlob`/`Must` checks the root folder for templates, and the second, `ParseGlob`/`Must`, checks any nested folders within the root for templates. Without the second, no templates in the nested folders will get pulled in.

We should be solid with no handling during the HTTP request since we moved the parsing of the user's templates during startup, so any errors would get triggered before the service started firing. If errors occur during the user template parsing, everything will revert to default before starting up HTTP processing.

DRuggeri commented on README.md:
```diff
@@ -182,6 +182,100 @@ curl http://127.0.0.1:8080/gotify_webhook -d '
+- User-defined templating allows matching and linking using the "define" name of the template [Go Templating](https://golang.google.cn/pkg/text/template/).
+- The Gotify software token is used for matching a message template.
+- The Gotify software token with `title=` at the beginning is used for matching a title template.
+- All file names must be unique but can be any name or subfolder naming you choose.
```
As noted in the prior comment, taking an opinion on expected file/directory structure will greatly simplify working with templates. This seems like adding flexibility where it is probably not needed - especially when it now means a user can create arbitrary directory structures, but has to be careful to never name two files the same.

What if it was just as simple as:
```
templates/
  common.tmpl
  apptoken1/
    title.tmpl
    message.tmpl
  apptoken2/
    title.tmpl
    subtemplate.tmpl
```
This structure makes it obvious which templates come into play for which tokens

IncognitoCoding replied:
My favorite part about seeing the power of Grafana was the complete customization they offer with Go Templating. My thoughts with the bridge were to allow the same full unrestricted customization for the bridge's templates.

For example, I may have a template for logging that I want to use for several different "applications" in Gotify. I can link this by using the defined variable quickly and efficiently without extra effort. The same could be said for title templates and broader message templates (networking, logging, etc.).

I see how the file structure could help users unfamiliar with Go Templating, but I suspect users unfamiliar with Go Templating would use the bridge's default messaging. My guess is that users using the bridge have Grafana installed and are familiar with the Go Templating in Grafana, which would allow them to use this functionality without any confusion. With a defined folder structure, the template would still need a defined variable (name) to match during the ExecuteTemplate function unless you have seen this handled differently.

DRuggeri replied:
Ah, yes - my thought is that since the go templating system [will default to the file name being used as the name of the template](https://pkg.go.dev/text/template#Template.ParseFiles) (as though it were `define`'ed in the template), we could rely on the file names being the template name unless decided otherwise by the user (placing a `define` inside). This would help a lot during template application since we can then use `template.Lookup("title")` during processing to get the title template. It also opens the door to the user to be as powerful as they'd like by defining many sub-templates (a feature I hadn't considered until I saw it in this PR) in various other files.

&nbsp;

> I see how the file structure could help users unfamiliar with Go Templating, but I suspect users unfamiliar with Go Templating would use the bridge's default messaging. My guess is that users using the bridge have Grafana installed and are familiar with the Go Templating in Grafana, which would allow them to use this functionality without any confusion

Interestingly enough, I don't use Grafana templates at all! I only use the [Prometheus alertmanager](https://prometheus.io/docs/alerting/latest/alertmanager/), which was the genesis of this particular project 😄 It wasn't until prior PRs discussed the Grafana templating capabilities that I even thought Grafana would be used as a source of alerts for the bridge.

&nbsp;
Hopefully the [longer comment](https://github.com/DRuggeri/alertmanager_gotify_bridge/pull/29#pullrequestreview-1304850857) I left explains the thought process. By loading as much up front as possible, supporting template overrides and customization via directory trees, and using shortcuts available thanks to template lookups by name, I think this will give the flexibility you describe while also being lean during message processing.

IncognitoCoding replied:
I did not realize Go Templating defaulted to the file name. That is good to know.

Good point on you not using Grafana templating. Since it is limited I am sure others have not even tried it yet. It has great potential if they allow it to alert with their alert engine when Prometheus alerts trigger. The limitation of Grafana is what allowed me to come across this bridge. I still believe people with an understanding of structuring single-file Go Templates (especially the functions), would be able to pick up on defining templates fairly quickly.

My original PR could load everything up front as well. I just had it loading during the handleCall for dynamic user template changes. The load would have no dependencies on what users call tokens are being sent. All user templates would load and if a token was sent during an alert, a matching token would utilize that template.

One issue I see with the folder `token` structure is special OS limitations on naming. Linux allows you to end a folder with a period, but Windows does not allow a folder to end with a period. The period would be stripped from any instance of the bridge running on Windows. Tokens are randomly generated with no ability to edit/change and could lead to issues with people implementing templating with existing applications. I have some applications in my Gotify that end with a period. We could try and add code to detect if a folder has 99% of the values, but we would be adding additional overhead to the bridge.

Few things from your high-level thought process:
1. The function example you listed would need some editing to handle sub-directories. Any root (common) templates would need to `Must` with the return *ut.Template as a variable, so that variable can pass to the ParseGlob for any sub-directory. That same would go for any looping sub/child directories. Without this, you would get "not defined" errors when referencing templates. That is why the "parseMsgTemplates" function I submitted in this PR had the `Walk` for the file structure before the ParseGlob.
2. Restricting the files to only tmpl would restrict users from using other popular file formats, that only syntax color in certain code editors (ex: VS Code).

DRuggeri replied:
> One issue I see with the folder token structure is special OS limitations on naming. Linux allows you to end a folder with a period, but Windows does not allow a folder to end with a period.

This is a good point. I had a suspicion that tokens may not be file or folder name safe - thanks for confirming. Indeed, the [characters used for tokens](https://github.com/gotify/server/blob/4c58f4a7423c61de3d0f4d64aa282d5eaea3d10a/auth/token.go#L9) are generally safe with the one exception you mention of Windows not permitting a folder ending in `.`... but who knows what else? Operating systems can be fun like that.

So, I guess the opinionated directory structure idea won't quite work out after all.


> My original PR could load everything up front as well. I just had it loading during the handleCall for dynamic user template changes. The load would have no dependencies on what users call tokens are being sent. All user templates would load and if a token was sent during an alert, a matching token would utilize that template.

Thanks, that helps explain the thought process. I like the idea of dynamic processing, but I would definitely prefer to load/fail up front so there are fewer opportunities to for gotchas during processing. How about both? Let's process all templates at the beginning and bail on any error (including the issues found in any file extension/etc), but discard the result. Then, during request handling, do it again in case the user has changed the files.

It does feel wasteful to process all the templates on every request for users that don't care for dynamic updates, but if that becomes an issue, I'm sure a new flag can be added down the line.

DRuggeri reviewed (changes requested):
This feels very complex and expects the user to understand go templating quite a bit. I've noted a few specific areas that I think need attention, but generally speaking, I think that this should be adjusted to do something like this...

## 
Specify an opinionated directory structure:
```
templates/
  common.tmpl
  token1/
    title.tmpl
    sub.tmpl
  token2/
    title.tmpl
    message.tmpl
```

This preferred layout will allow for efficient parsing and use - see below

##
During startup, parsing templates like so would be trivial - and would allow catching failures quickly

```go
// in main function - init
userTemplates, err := parseUserTemplates("somePath")
if err != nil {
  //panic...
}

// Parses common templates and allows for per-token overrides of templates. Results are stored in a map of templates keyed on the app token
func parseUserTemplates(baseDir string) (map[string]*ut.Template, error) {
  ret := make(map[string]*Template)

  //get list of base or common templates
  common, err := ut.ParseGlob(fmt.Sprintf("%s%c*.tmpl", baseDir, filepath.Separator))
  if err != nil {
    return ret, err
  }

  //in a loop through child directories in baseDir, assuming token is the dirname
    subdir := fmt.Sprintf("%s%c%s", baseDir, filepath.Separator, token)
    tokenTemplates := common.Clone()
    tokenTemplates, err = ut.ParseGlob(fmt.Sprintf("%s%c*.tmpl", subdir, filepath.Separator))
    //Check for errors and return if needed
    ret[token] = tokenTemplates

  return ret, nil
}
```

##
Later on, when attempting to use the templates, we can take shortcuts:
```go
// during usage of the templates
var tmpl *ut.Template

if temp, ok := userTemplates[token]; ok {
  if temp = temp.Lookup("title"); temp != nil {
    tmpl = temp
  }
}
if tmpl == nil {
  //Try to create template from the annotation, bail if we cannot
}

// render with the template as already done
```

IncognitoCoding committed:
Moved user tmpl parse to the main, grouped some code, and renamed a few functions.

1. Moved the user template parsing from the handleCall to the main.
2. Grouped the title and message user template execution.
3. Renamed functions for a cleaner read.
4. Removed HTML as a template extension and updated the ReadMe.

IncognitoCoding commented:
I had some spare time this evening, so I revised the code with all discussed changes with the exception of the folder structure with tokens since that piece is still being discussed. 

### Changes:
1. Moved the user template parsing from the handleCall to the main.
2. Grouped the title and message user template execution.
3. Renamed functions for a cleaner read.
4. Removed HTML as a template extension and updated the ReadMe.

IncognitoCoding committed:
Moved the defaultTitle and defaultMsg.

1. The defaultTitle and defaultMsg flags needed to be moved to prevent panic if a user-defined template rendering failed.

DRuggeri commented:
Great - thanks! Apologies for the delay in updates, I was out of pocket the past week or so.

IncognitoCoding committed:
Update README.md

Update to the "Usage Notes" for templating.

IncognitoCoding commented:
> Great - thanks! Apologies for the delay in updates, I was out of pocket the past week or so.

No worries. Life can sometimes get in the way of hobbies.

v4rakh commented:
Would that solve my question in #6 ?

IncognitoCoding commented:
> Would that solve my question in #6 ?

Yes, this pull request (PR) will allow you to create one or multiple user-defined templates to send customized messages to Gotify. You can use any key (including: `status`, `generatorURL`, etc.) from the JSON sent from alertmanager. Depending on the product, some alerts may have more detail outside the annotations section, which user-defined templating would allow any usage of those keys/values.

**You would have two options:**
1. Create a single template with `if/else` to set the status of `firing` or `resolved`  based on the received `status`.
2. Create customized `Firing` and `Resolved` templates.

You can read more details on the usage in this PR's ReadMe file under the `Bridge Message Templating` section.

v4rakh commented:
Would love to see that being incorporated.

DRuggeri commented:
Thanks so much for accepting and incorporating the feedback! I'll go ahead and release this version today!

v4rakh commented:
Thanks for releasing it so quickly.

I am not able to get it working properly, maybe I miss something here. I like to have a very simple icon showing when  alerts fire or are resolved, together with the actual title, then the description.

I've successfully mounted to `/./templates` (using docker, should we document that this is root in case of the docker image or allow setting a path/env var?) and it's loading the templates (at least no error which I had before).

```shell
{{ define "MyGotifyToken" }}
{{ if eq (.Status) ("firing") }}🔥 {{else}}✅ {{end}}
{{ .Annotations.summary }}
{{ .GeneratorURL }}
{{end}}
```

I am sending the following request to my `/gotify_webhook` url:

```shell
{ "alerts": [
  {
    "annotations": {
      "description":"A description",
      "summary":"hmmm test",
      "priority":"critical"
    },
    "status": "firing",
    "generatorURL": "http://foobar.com"
  }
]}
```

I am not getting any icon, neither a generator URL.

Could you maybe outline a simple working example?

IncognitoCoding commented:
> Thanks for releasing it so quickly.
> 
> I am not able to get it working properly, maybe I miss something here. I like to have a very simple icon showing when alerts fire or are resolved, together with the actual title, then the description.
> 
> I've successfully mounted to `/./templates` (using docker, should we document that this is root in case of the docker image or allow setting a path/env var?) and it's loading the templates (at least no error which I had before).
> 
> ```shell
> {{ define "MyGotifyToken" }}
> {{ if eq (.Status) ("firing") }}🔥 {{else}}✅ {{end}}
> {{ .Annotations.summary }}
> {{ .GeneratorURL }}
> {{end}}
> ```
> 
> I am sending the following request to my `/gotify_webhook` url:
> 
> ```shell
> { "alerts": [
>   {
>     "annotations": {
>       "description":"A description",
>       "summary":"hmmm test",
>       "priority":"critical"
>     },
>     "status": "firing",
>     "generatorURL": "http://foobar.com"
>   }
> ]}
> ```
> 
> I am not getting any icon, neither a generator URL.
> 
> Could you maybe outline a simple working example?

Using the bind mount with `/./templates` is the correct way to mount your templates folder. I will make an edit to the `ReadMe`.

- Bind Example: `../alertmanager_gotify_bridge/templates:/./templates`

Your test case is correct as long as you have entered the correct `MyGotifyToken`. 

Since you are not getting an `icon` or `generator URL`, I assume you updated or set up your testing templates after the bridge was already running. All user-templates are loaded during startup. Any template changes will always require a restart or update to the service to see the changes reflect. Please restart or update the service and try again.

v4rakh commented:
@IncognitoCoding Yes, I have restarted the service, actually I did `down` and `up` with `docker-compose`, so that should be all OK. My gotify token is also correct, it's the same as I've used as in `GOTIFY_TOKEN=`.

It doesn't seem to match. Doing a POST request to the `/gotify_webhook` should always use that token, right, so not sure. It should be working.

IncognitoCoding commented:
> @IncognitoCoding Yes, I have restarted the service, actually I did `down` and `up` with `docker-compose`, so that should be all OK. My gotify token is also correct, it's the same as I've used as in `GOTIFY_TOKEN=`.
> 
> It doesn't seem to match. Doing a POST request to the `/gotify_webhook` should always use that token, right, so not sure. It should be working.

I figured out your issue. The problem is a variable being passed during the template process. A typo in the code had the `token override` value as the only option to use the templates. You are likely using the default `GOTIFY_TOKEN` set in the docker. This was a simple fix, and I am submitting a PR request for this fix today.

As a temporary workaround, use the token override in your URL. You can also keep the token you have set in your docker file. The token override will override any token in the docker-compose file. Once the PR is approved,  your template will work with just the docker token.
- Token Override Example: http://127.0.0.1:8080/gotify_webhook?token=MyGotifyToken

v4rakh commented:
Thanks. Works when passing via query parameter.
